linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala

```diff
@@ -41,50 +40,120 @@ private[emitter] final class VarGen(jsGen: JSGen, nameGen: NameGen,
+    foldSameModule(scope) {
+      VarRef(ident)
+    } { moduleID =>
+      DotSelect(VarRef(internalModuleFieldIdent(moduleID)), ident)
+    }
   }
 
   def globalClassDef[T: Scope](field: String, scope: T,
       parentClass: Option[Tree], members: List[Tree],
       origName: OriginalName = NoOriginalName)(
-      implicit pos: Position): Tree = {
+      implicit moduleContext: ModuleContext, pos: Position): WithGlobals[Tree] = {
     val ident = globalVarIdent(field, scope, origName)
-    ClassDef(Some(ident), parentClass, members)
+    maybeExport(ident, ClassDef(Some(ident), parentClass, members), mutable = false)
   }
 
   def globalFunctionDef[T: Scope](field: String, scope: T,
       args: List[ParamDef], body: Tree,
       origName: OriginalName = NoOriginalName)(
-      implicit pos: Position): Tree = {
-    FunctionDef(globalVarIdent(field, scope, origName), args, body)
+      implicit moduleContext: ModuleContext, pos: Position): WithGlobals[Tree] = {
+    val ident = globalVarIdent(field, scope, origName)
+    maybeExport(ident, FunctionDef(ident, args, body), mutable = false)
   }
 
   def globalVarDef[T: Scope](field: String, scope: T, value: Tree,
-      origName: OriginalName = NoOriginalName, mutable: Boolean = false)(
-      implicit pos: Position): Tree = {
-    genLet(globalVarIdent(field, scope, origName), mutable, value)
+      origName: OriginalName = NoOriginalName)(
+      implicit moduleContext: ModuleContext, pos: Position): WithGlobals[Tree] = {
+    val ident = globalVarIdent(field, scope, origName)
+    maybeExport(ident, genConst(ident, value), mutable = false)
   }
 
+  /** Attention: A globalVarDecl may only be modified from the module it was declared in. */
   def globalVarDecl[T: Scope](field: String, scope: T,
       origName: OriginalName = NoOriginalName)(
-      implicit pos: Position): Tree = {
-    genEmptyMutableLet(globalVarIdent(field, scope, origName))
+      implicit moduleContext: ModuleContext, pos: Position): WithGlobals[Tree] = {
+    val ident = globalVarIdent(field, scope, origName)
+    maybeExport(ident, genEmptyMutableLet(ident), mutable = true)
   }
 
-  // Still public for field exports.
-  def globalVarIdent[T](field: String, scope: T,
+  /** Unlike a mutable VarDecl, a globallyMutableVarDef may be modified from any
+   *  module. As such, an additional field needs to be provided for an
+   *  additional setter. This is used when generating ES modules.
+   */
+  def globallyMutableVarDef[T: Scope](field: String, setterField: String,
+      scope: T, value: Tree, origName: OriginalName = NoOriginalName)(
+      implicit moduleContext: ModuleContext, pos: Position): WithGlobals[Tree] = {
+    val ident = globalVarIdent(field, scope, origName)
+    val varDef = genLet(ident, mutable = true, value)
+
+    if (config.moduleKind == ModuleKind.ESModule && !moduleContext.public) {
+      val setterIdent = globalVarIdent(setterField, scope, origName)
```

Comment:
I don't think it makes sense to give the original name of the field to the setter. They are separate entities.